File: cpp/0103_zigzagLevelOrder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// DFS: 无脑DFS。。奇数层reverse
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        traverse(root,res,0);
        if(res.size()<2)
            return res;
        for(auto i=1;i<res.size();i+=2)
            reverse(res[i].begin(),res[i].end());
        return res;
    }

    void traverse(TreeNode* root,vector<vector<int>> &res,int depth)
    {
        if(root==nullptr)
            return;
        if(res.size()>depth)
            res[depth].push_back(root->val);
        else
            res.push_back({root->val});
        traverse(root->left,res,depth+1);
        traverse(root->right,res,depth+1);
    }
};

File: java/0103_zigzagLevelOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        traverse(root,res,0);
        return res;
    }
    public void traverse(TreeNode root,List<List<Integer>> res,int depth)
    {
        if(root==null)
            return;
        if(res.size()>depth)
        {
            if(depth%2==0)
                res.get(depth).add(root.val);
            else
                res.get(depth).add(0,root.val);
        }
        else
        {
            if(depth%2==0)
                res.add(new ArrayList<>(Arrays.asList(root.val)));
            else
                res.add(new LinkedList<>(Arrays.asList(root.val)));
        }
            
        traverse(root.left,res,depth+1);
        traverse(root.right,res,depth+1);
    }
}
